### PR TITLE
Add ResponseErrorHandler to handle REST calls

### PR DESCRIPTION
**Pull request by javierscope**

### CODE REVIEW

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	@Override
+	public void handleError(ClientHttpResponse clientHttpResponse) throws IOException {
+		String responseAsString = toString(clientHttpResponse.getBody());
+		log.error("ResponseBody: {}", responseAsString);
+
+		throw new CustomException(responseAsString);
+	}
+
+	@Override
+	public void handleError(URI url, HttpMethod method, ClientHttpResponse response) throws IOException {
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:53]: if (httpResponse.getStatusCode()
          .series() == HttpStatus.Series.SERVER_ERROR) {
             // log its a server error 
        } else if (httpResponse.getStatusCode()
          .series() == HttpStatus.Series.CLIENT_ERROR) {
             // log its client error
            if (httpResponse.getStatusCode() == HttpStatus.NOT_FOUND) {
                // log not found error
                throw new NotFoundException();
            }
        }

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+		String responseAsString = toString(response.getBody());
+		log.error("URL: {}, HttpMethod: {}, ResponseBody: {}", url, method, responseAsString);
+
+		throw new CustomException(responseAsString);
+	}
+
+	String toString(InputStream inputStream) {
+		Scanner s = new Scanner(inputStream).useDelimiter("\\A");
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:61]: Aren't we printing entire failed response body?

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =====
+ */
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class CustomResponseErrorHandler implements ResponseErrorHandler {
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:36]: Rename to GenericRestResponseErrorHandler

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	@Override
+	public boolean hasError(ClientHttpResponse clientHttpResponse) throws IOException {
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:39]: Add test cases around this

File: src/main/java/com/scoperetail/fusion/core/adapter/out/web/http/impl/PosterOutboundHttpAdapterImpl.java [hunk 5]
```diff
@@ -47,6 +49,7 @@ public void post(final String url, final String methodType, final String request
     HttpComponentsClientHttpRequestFactory clientHttpRequestFactory =
         new HttpComponentsClientHttpRequestFactory(HttpClientBuilder.create().build());
     final RestTemplate restTemplate = new RestTemplate(clientHttpRequestFactory);
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/adapter/out/web/http/impl/PosterOutboundHttpAdapterImpl.java:51]: Also add ClientHttpRequestInterceptor - to log the request details before sending the request

**Review by amit-petkar — CHANGES_REQUESTED** [review]
Need changes.

File: src/main/java/com/scoperetail/fusion/core/adapter/out/web/http/impl/PosterOutboundHttpAdapterImpl.java (hunk 5 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/adapter/out/web/http/impl/PosterOutboundHttpAdapterImpl.java:51]: We are already printing the requests, do we still need the Interceptor?

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java (hunk 3 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:36]: Ok

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java (hunk 1 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:53]: Ok

File: src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java (hunk 2 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/CustomResponseErrorHandler.java:61]: Yes

File: src/main/java/com/scoperetail/fusion/core/adapter/out/web/http/impl/PosterOutboundHttpAdapterImpl.java (hunk 5 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/adapter/out/web/http/impl/PosterOutboundHttpAdapterImpl.java:53]: Loggin interceptor added

File: src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =====
+ */
+
+import lombok.extern.slf4j.Slf4j;
+import org.webjars.NotFoundException;
+
+@Slf4j
+public class GenericRestResponseErrorHandler implements ResponseErrorHandler {
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java:37]: Spring already has DefaultResponseErrorHandler. Are we doing anything different?

File: src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+			log.error("Not Found Error. URL: {}, HttpMethod: {}, ResponseBody: {}", url, method, responseAsString);
+			throw new NotFoundException(responseAsString);
+		} else {
+			log.error("Generic Error. URL: {}, HttpMethod: {}, ResponseBody: {}", url, method, responseAsString);
+			throw new CustomException(responseAsString);
+		}
+	}
+
+	String toString(InputStream inputStream) {
+		Scanner s = new Scanner(inputStream).useDelimiter("\\A");
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java:70]: Why just a single line?

File: src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+		return s.hasNext() ? s.next() : "";
+	}
+
+	static class CustomException extends IOException {
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java:74]: Any specific need to create custom exception?

File: src/main/java/com/scoperetail/fusion/core/common/LoggingInterceptor.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.http.client.ClientHttpResponse;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class LoggingInterceptor implements ClientHttpRequestInterceptor {
+
+	@Override
+	public ClientHttpResponse intercept(HttpRequest req, byte[] reqBody, ClientHttpRequestExecution ex)
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/LoggingInterceptor.java:40]: https://howtodoinjava.com/spring-boot2/resttemplate/clienthttprequestinterceptor/

**Review by amit-petkar — CHANGES_REQUESTED** [review]
Need chages.

File: src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java (hunk 7 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java:70]: This is reading all the message

File: src/main/java/com/scoperetail/fusion/core/common/LoggingInterceptor.java (hunk 9 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/LoggingInterceptor.java:40]: logging request and response as per comment

File: src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java (hunk 6 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java]: using default handler

File: src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java (hunk 8 above)
Comment (javierscope) [reply src/main/java/com/scoperetail/fusion/core/common/GenericRestResponseErrorHandler.java]: custom exception removed

File: src/main/java/com/scoperetail/fusion/core/common/LoggingInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	@Override
+	public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution)
+			throws IOException {
+
+		logRequest(request, body);
+		ClientHttpResponse response = execution.execute(request, body);
+		logResponse(response);
+
+		//Add optional additional headers
+		response.getHeaders().add("headerName", "VALUE");
```
Comment (amit-petkar) [inline src/main/java/com/scoperetail/fusion/core/common/LoggingInterceptor.java:53]: ???

**Review by amit-petkar — CHANGES_REQUESTED** [review]
need changes.

**Review by amit-petkar — APPROVED** [review]
Approved.

**Pull request merged**